disnake/iterators.py

```diff
@@ -398,6 +400,103 @@ async def _retrieve_messages_around_strategy(self, retrieve):
         return []
 
 
+class BanIterator(_AsyncIterator["BanEntry"]):
+    """Iterator for receiving a guild's bans.
+
+    The bans endpoint has two behaviours we care about here:
+    If ``before`` is specified, the bans endpoint returns the ``limit``
+    bans with user ids before ``before``, sorted with smallest first. For filling over
+    1000 bans, update the ``before`` parameter to the largest user id received.
+    If ``after`` is specified, it returns the ``limit`` bans with user ids after
+    ``after``, sorted with smallest first. For filling over 1000 bans, update the
+    ``after`` parameter to the smallest user id received.
+
+    A note that if both ``before`` and ``after`` are specified, ``after`` is ignored by the
+    bans endpoint.
+
+    Parameters
+    -----------
+    guild: :class:`~disnake.Guild`
+        The guild to get bans from.
+    limit: Optional[:class:`int`]
+        Maximum number of bans to retrieve.
+    before: Optional[:class:`abc.Snowflake`]
+        Date or user id before which all bans must be.
+    after: Optional[:class:`abc.Snowflake`]
+        Date or user id after which all bans must be.
+    """
+
+    def __init__(
+        self,
+        guild: Guild,
+        limit: Optional[int] = None,
+        before: Optional[Snowflake] = None,
+        after: Optional[Snowflake] = None,
+    ):
+        self.guild = guild
+        self.limit = limit
+        self.before = before
+        self.after = after or OLDEST_OBJECT
+
+        self.state = self.guild._state
+        self.get_bans = self.state.http.get_bans
+        self.bans = asyncio.Queue()
+
+        if self.before:
```

Comment:
This should manually filter results in case both `before` and `after` but no `limit` is given, similar to `GuildIterator`.
I'll avoid making comments all over the place and instead just propose this diff:
```diff
diff --git a/disnake/iterators.py b/disnake/iterators.py
index a6425667..d317dda5 100644
--- a/disnake/iterators.py
+++ b/disnake/iterators.py
@@ -442,8 +442,11 @@ class BanIterator(_AsyncIterator["BanEntry"]):
         self.get_bans = self.state.http.get_bans
         self.bans = asyncio.Queue()

+        self._filter: Optional[Callable[[BanPayload], bool]] = None
         if self.before:
             self._retrieve_bans = self._retrieve_bans_before_strategy
+            if self.after != OLDEST_OBJECT:
+                self._filter = lambda b: int(b["user"]["id"]) > self.after.id
         else:
             self._retrieve_bans = self._retrieve_bans_after_strategy

@@ -468,6 +471,9 @@ class BanIterator(_AsyncIterator["BanEntry"]):
             if len(data) < 1000:
                 self.limit = 0  # terminate the infinite loop

+            if self._filter:
+                data = filter(self._filter, data)
+
             for element in data:
                 await self.bans.put(
                     BanEntry(
```